### ISSUE TITLE
Engine roll failures

### ISSUE THREAD

**Issue by zanderso** [flutter/flutter#88205]
The Engine roll here https://github.com/flutter/flutter/pull/88186 encountered multiple different failures on framework CI.

It looks like several iOS test apps failed to start, or crashed shortly after starting: https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20integration_test_test_ios/1471/overview

There was also a failure in an accessibility test on Android

https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_android%20android_semantics_integration_test/2203/overview

@chunhtai @a-siva @iskakaushik

Speculatively marking `dependency: dart` since this roll contained the compressed pointer change from the VM.

**Comment by jason-simmons** [comment]
Confirmed that the iOS issue started with the Dart compressed pointer change.

With a local roll to https://github.com/flutter/engine/commit/d856fc33baf1a93feac8aa2369de1d139907bc4e `flutter run` of a default template app crashes on iOS:
```
(lldb) ../../third_party/dart/runtime/vm/virtual_memory_posix.cc: 90: error: Failed to reserve region for compressed heap: 12 (Cannot allocate memory)
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x1fb3)
    frame #0: 0x0000000100da4f38 Flutter`dart::Profiler::DumpStackTrace(unsigned long, unsigned long, unsigned long, bool) + 112
Flutter`dart::Profiler::DumpStackTrace:
->  0x100da4f38 <+112>: ldrb   w8, [x0, #0x8]
    0x100da4f3c <+116>: cbz    w8, 0x100da4f58           ; <+144>
    0x100da4f40 <+120>: cbz    x23, 0x100da4f68          ; <+160>
    0x100da4f44 <+124>: ldr    x0, [x19]
```

**Comment by zanderso** [comment]
Reverts have landed and built, and rolls have restarted.

**Comment by a-siva** [comment]
Thanks for trying this out locally and figuring out the error message @jason-simmons .

Question regarding mmap on iOS, is there an inherent limit on the amount of memory that can be reserved using a mmap call on iOS. We are trying to reserve 4 GB of memory here and it appears to be failing, I presume there is some kind of limit.

**Comment by zanderso** [comment]
Maybe @jmagman or @gaaclarke might know.

**Comment by gaaclarke** [comment]
It sounds like it can't find an appropriate region in memory for 4GB.  For error 12 (ENOMEM) the documentation says:
```
     [ENOMEM]           MAP_FIXED was specified and the addr argument was not available.  MAP_FIXED was specified and the address range specified exceeds the
                        address space limit for the process.  MAP_ANON was specified and insufficient memory was available.
```

I saw some old documentation saying that 700MB was a limit and referenced an old blog post by john carmack I couldn't find.  The advice given was the break it up into chunks instead of a large region.

**Comment by a-siva** [comment]
> I saw some old comments saying that 700MB was a limit and referenced an old blog post by john carmack I couldn't find. The advice given was the break it up into chunks instead of a large region.

For compressed pointers to work we need a contiguous region of memory and hence this reservation upfront, breaking it up into smaller chunks would mean we don't get a contiguous region.

**Comment by chinmaygarde** [comment]
I don't know what the consequences are of allocating a region that large or how the VM manages these allocation or the pages/slabs within it. But, a couple of things to try to get a contiguous allocation that large:

* Use `vm_allocate` to allocate the virtual memory. This should only be restricted to the available address space. Presumably, you'll run into issues only when a page fault on access cannot be backed by physical memory. Again, I don't know how the VM manages to avoid this on other platforms but at least you'll get a contiguous arena.
* If `MAP_ANONYMOUS` is the issue when using `mmap`, you could just not use anonymous mappings by creating a temporary FD and using that for the file mapping. Something like `mkstemp` -> `ftruncate` -> `mmap(... MAP_FILE ...)` -> `close()`.

Again, I don't know how the VM manages to avoid too many pages resident in physical memory when creating the contiguous allocation. But I expect it's possible to allocate virtual all the way up to address space exhaustion 🤷🏽 .
